002-es-modules.md

```diff
@@ -0,0 +1,525 @@
+//   ./node_modules/abc/123/package.json
+//   ./node_modules/abc/123/index.mjs
+//   ./node_modules/abc/123/index.js
+// and parent node_modules:
+//   ../node_modules/abc/123.mjs
+//   ../node_modules/abc/123.js
+//   ../node_modules/abc/123/package.json
+//   ../node_modules/abc/123/index.mjs
+//   ../node_modules/abc/123/index.js
+//   etc.
+import 'abc/123';
+```
+
+#### 3.3.1. Removal of non-local dependencies
+
+All of the following will not be supported by the `import` statement:
+
+* `$NODE_PATH`
+* `$HOME/.node_modules`
+* `$HOME/.node_libraries`
+* `$PREFIX/lib/node`
+* `module/../../outside-of-module`
+
+Use local dependencies, and symbolic links as needed.
+
+##### 3.3.1.1. How to support non-local dependencies
+
+Although not recommended, and in fact discouraged, there is a way to support
+non-local dependencies. **USE THIS AT YOUR OWN DISCRETION**.
+
+Symlinks of `node_modules -> $HOME/.node_modules`, `node_modules/foo/ ->
+$HOME/.node_modules/foo/`, etc. will continue to be supported.
+
+Adding a parent directory with `node_modules` symlinked will be an effective
+strategy for recreating these functionalities. This will incur the known
+problems with non-local dependencies, but now leaves the problems in the hands
+of the user, allowing Node to give more clear insight to your modules by
+reducing complexity.
+
+Given:
+
+```sh
+/opt/local/myapp
+```
+
+Transform to:
+
+```sh
+/opt/local/non-local-deps/myapp
+/opt/local/non-local-deps/node_modules -> $PREFIX/lib/node (etc.)
+```
+
+And nest as many times as needed.
+
+#### 3.3.2. Errors from new path behavior.
+
+In the case that an `import` is unable to find a module, Node should
+make a **best effort** to see if `require` would have found the module and
+print out where it was found, if `NODE_PATH` was used, if `HOME` was used, etc.
+
+#### 3.3.3. Shipping both ESM and NCJS
+
+When a `package.json` main is encountered, file extension searches are used to
+provide a means to ship both ESM and NCJS variants of packages. If we have two
+entry points `index.mjs` and `index.js` setting `"main":"./index"` in
+`package.json` will make Node pick up either, depending on what is supported.
+
+##### 3.3.3.1. Excluding main
+
+Since `main` in `package.json` is entirely optional even inside of npm
+packages, some people may prefer to exclude main entirely in the case of using
+`./index` as that is still in the Node module search algorithm.
+
+### 3.4. ESM Evaluation
+
+#### 3.4.1. Environment Variables
+
+ESM will not be bootstrapped with magic variables and will await upcoming specifications in order to provide such behaviors in a standard way. As such, the following variables are changed:
+
+| Variable | Exists | Value |
+| ---- | ---- | --- |
+| this | y | [`undefined`](https://tc39.github.io/ecma262/#sec-module-environment-records-getthisbinding) |
+| arguments | n | |
+| require | n | |
+| module | n | |
+| exports | n | |
+| __filename | n | |
+| __dirname | n | |
```

Comment:
Ah, I misunderstood; they will actually not exist anymore and thus should become an attempt to read against `global` (which will probably result in `undefined`).

I suppose that, until they are specified, a workaround to, at least, have _some_ known dirname, would be to have a NCJS module that just does `module.exports = __dirname`.